### PR TITLE
BUG: asymmetric error bars for series (GH9536)

### LINKED ISSUE #9536: Asymmetric error bars are not supported for series (only dataframes)
According to [the documentation](http://pandas-docs.github.io/pandas-docs-travis/visualization.html#plotting-with-error-bars), asymmetric error bars should be able to be passed in to `plot` as an Mx2 array for a `Series` object or a Mx2xN array for a `DataFrame` object.

However, the code only seems to parse this correctly for a `DataFrame`: https://github.com/pydata/pandas/blob/master/pandas/tools/plotting.py#L1338

When passing raw error values as `yerr`, it seems that the first values are used as two-sided error differences, instead of being plotted as raw upper and lower values. Note the symmetric error bars below:

image

Using the underlying matplotlib function directly yields the expected results, but without the pandas-generated decorations:

image

This was originally noted in [this StackOverflow question](http://stackoverflow.com/q/26793758/586086) and I corroborated the other user's experience.

### PR DESCRIPTION
**Pull request by Steffen911**
closes #9536

This fix enables asymmetric errors bars for pd.Series plots. In the current implementation the bars are symmetric, even if two sets of errors are provided.

Took inspirations from #12046

### CODE REVIEW

File: pandas/tests/plotting/test_series.py [hunk 1]
```diff
@@ -729,6 +729,22 @@ def test_dup_datetime_index_plot(self):
...
+        s = Series(np.arange(10), name="x")
+        err = np.random.rand(2, 10)
+
+        ax = s.plot(yerr=err, xerr=err)
+
+        yerr_0_0 = ax.collections[1].get_paths()[0].vertices[:, 1]
+        expected_0_0 = err[:, 0] * np.array([-1, 1])
+        tm.assert_almost_equal(yerr_0_0, expected_0_0)
+
+        with pytest.raises(ValueError):
```
Comment (phofl) [inline pandas/tests/plotting/test_series.py:743]: Could you please check the error message too?

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -288,6 +288,7 @@ Other enhancements
 - :meth:`HDFStore.put` now accepts `track_times` parameter. Parameter is passed to ``create_table`` method of ``PyTables`` (:issue:`32682`).
 - Make :class:`pandas.core.window.Rolling` and :class:`pandas.core.window.Expanding` iterable（:issue:`11704`)
 - Make ``option_context`` a :class:`contextlib.ContextDecorator`, which allows it to be used as a decorator over an entire function (:issue:`34253`).
+- ``Series.plot`` fix asymmetric error bars (:issue:`9536`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.1.0.rst:291]: Could you described the previous and the current behavior to make the fix a bit more clear?

File: pandas/tests/plotting/test_series.py (hunk 1 above)
Comment (Steffen911) [reply pandas/tests/plotting/test_series.py]: Added in https://github.com/pandas-dev/pandas/pull/34514/commits/99b51fc91279a486b9fe56eab6df6f93f40d292f and https://github.com/pandas-dev/pandas/pull/34514/commits/0ea50d35ec302c0015c9cabedb77b8b91b1a41df

File: doc/source/whatsnew/v1.1.0.rst (hunk 2 above)
Comment (Steffen911) [reply doc/source/whatsnew/v1.1.0.rst]: I extended the description in https://github.com/pandas-dev/pandas/pull/34514/commits/4a2eb9cbb4753ed6b1fc2c9e14fe4ef3b4fc6f18

File: pandas/tests/plotting/test_series.py [hunk 3]
```diff
@@ -729,6 +729,26 @@ def test_dup_datetime_index_plot(self):
...
 
+    def test_errorbar_asymmetrical(self):
+        # GH9536
+        s = Series(np.arange(10), name="x")
+        err = np.random.rand(2, 10)
+
+        ax = s.plot(yerr=err, xerr=err)
+
+        yerr_0_0 = ax.collections[1].get_paths()[0].vertices[:, 1]
+        expected_0_0 = err[:, 0] * np.array([-1, 1])
```
Comment (MarcoGorelli) [inline pandas/tests/plotting/test_series.py:740]: Would it be possible to check all the points here? Something like
```python
result = np.vstack([i.vertices[:, 1] for i in ax.collections[1].get_paths()])
expected = err.T * np.array([-1, 1])
```
?

File: doc/source/whatsnew/v1.1.0.rst [hunk 4]
```diff
@@ -289,6 +289,9 @@ Other enhancements
 - Make :class:`pandas.core.window.Rolling` and :class:`pandas.core.window.Expanding` iterable（:issue:`11704`)
 - Make ``option_context`` a :class:`contextlib.ContextDecorator`, which allows it to be used as a decorator over an entire function (:issue:`34253`).
 - :meth:`groupby.transform` now allows ``func`` to be ``pad``, ``backfill`` and ``cumcount`` (:issue:`31269`).
+- :meth:`Series.plot` now supports asymmetric error bars. Previously, if :meth:`Series.plot` received a "2xN" array with
+  error values for `yerr` and/or `xerr`, the left/lower values (first row) would be mirrored, while the right/upper values (second row) are
+  ignored. Now, the first row represents the left/lower error values and the second row the right/upper error values. (:issue:`9536`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v1.1.0.rst:294]: Couple of things:

1. Can you put this all on a single line? It's fine if it's long
2. Can you make the tenses agree? So,
```diff
- while the right/upper values (second row) are ignored.
+ while the right/upper values (second row) were ignored.
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks @Steffen911

File: pandas/tests/plotting/test_series.py (hunk 3 above)
Comment (Steffen911) [reply pandas/tests/plotting/test_series.py]: Addressed in https://github.com/pandas-dev/pandas/pull/34514/commits/2975406501f8942ded75091594f094e8d3095a2a

File: doc/source/whatsnew/v1.1.0.rst (hunk 4 above)
Comment (Steffen911) [reply doc/source/whatsnew/v1.1.0.rst]: Updated in https://github.com/pandas-dev/pandas/pull/34514/commits/fa4915b7e39251b6ad6d19cdd6355009f94bce80

File: pandas/tests/plotting/test_series.py [hunk 5]
```diff
@@ -729,6 +729,26 @@ def test_dup_datetime_index_plot(self):
...
+    def test_errorbar_asymmetrical(self):
+        # GH9536
+        s = Series(np.arange(10), name="x")
+        err = np.random.rand(2, 10)
+
+        ax = s.plot(yerr=err, xerr=err)
+
+        result = np.vstack([i.vertices[:, 1] for i in ax.collections[1].get_paths()])
+        expected = (err.T * np.array([-1, 1])) + list(zip(s, s))
+        tm.assert_almost_equal(result, expected)
```
Comment (MarcoGorelli) [inline pandas/tests/plotting/test_series.py:741]: I think this should be `tm.assert_numpy_array_equal`

File: pandas/tests/plotting/test_series.py [hunk 6]
```diff
@@ -729,6 +729,26 @@ def test_dup_datetime_index_plot(self):
...
         _check_plot_works(s.plot)
 
+    def test_errorbar_asymmetrical(self):
+        # GH9536
+        s = Series(np.arange(10), name="x")
+        err = np.random.rand(2, 10)
+
+        ax = s.plot(yerr=err, xerr=err)
+
+        result = np.vstack([i.vertices[:, 1] for i in ax.collections[1].get_paths()])
```
Comment (MarcoGorelli) [inline pandas/tests/plotting/test_series.py:739]: IMO the last operand would clearer as
```python
s.to_numpy().reshape(-1, 1)
```
, but it's up to you

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
@datapythonista do you want to have a look?

File: pandas/tests/plotting/test_series.py [hunk 7]
```diff
@@ -729,6 +729,26 @@ def test_dup_datetime_index_plot(self):
...
+        expected = (err.T * np.array([-1, 1])) + list(zip(s, s))
```
Comment (MarcoGorelli) [inline pandas/tests/plotting/test_series.py:740]: IMO the last operand would clearer as
```python
s.to_numpy().reshape(-1, 1)
```
, but it's up to you

File: pandas/tests/plotting/test_series.py (hunk 5 above)
Comment (Steffen911) [reply pandas/tests/plotting/test_series.py]: You're right. I had some floating point issues in the beginning, but now the numbers seem to be stable. Changed in https://github.com/pandas-dev/pandas/pull/34514/commits/4f430b3feafc948875783f847c0ba0a1d64692e1

File: pandas/tests/plotting/test_series.py (hunk 7 above)
Comment (Steffen911) [reply pandas/tests/plotting/test_series.py]: Updated in https://github.com/pandas-dev/pandas/pull/34514/commits/67b534d26acd5c722e561c3739a4ee91a6550736

**Comment by Steffen911** [comment]
@MarcoGorelli @datapythonista Could one of you review the PR? Or maybe suggest another reviewer?

File: pandas/plotting/_matplotlib/core.py [hunk 8]
```diff
@@ -795,7 +801,15 @@ def match_labels(data, e):
...
-            if err.ndim == 3:
+            if isinstance(self.data, ABCSeries) and err_shape[0] == 2:
+                err = np.expand_dims(err, 0)
+                err_shape = err.shape
+                if err_shape[2] != len(self.data):
+                    raise ValueError(
+                        "Asymmetrical error bars should be provided "
+                        f"with the shape (2, {len(self.data)})"
+                    )
+            if isinstance(self.data, ABCDataFrame) and err.ndim == 3:
```
Comment (datapythonista) [inline pandas/plotting/_matplotlib/core.py:812]: Noe important, but I think an `elif` makes more sense.
```suggestion
            elif isinstance(self.data, ABCDataFrame) and err.ndim == 3:
```

**Review by datapythonista — APPROVED** [review]
lgtm, thanks @Steffen911

File: pandas/plotting/_matplotlib/core.py [hunk 9]
```diff
@@ -810,7 +816,15 @@ def match_labels(data, e):
             err_shape = err.shape
 
             # asymmetrical error bars
-            if err.ndim == 3:
+            if isinstance(self.data, ABCSeries) and err_shape[0] == 2:
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/core.py:819]: Can you just have one branch that works generically for Series / DataFrame?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks pretty good minor comment

**Review by jreback — COMMENTED** [review]
lgtm ex @WillAyd comments.

File: pandas/plotting/_matplotlib/core.py (hunk 9 above)
Comment (Steffen911) [reply pandas/plotting/_matplotlib/core.py:819]: I don't think it's possible or useful. The error message is similar, but we perform some additional actions for the Series (adding a dimension) and updating the err_shape.

Comment (WillAyd) [reply pandas/plotting/_matplotlib/core.py:819]: OK. @jreback feel free to merge

**Comment by jreback** [comment]
lgtm can u merge master and ping on green

File: doc/source/whatsnew/v1.1.0.rst [hunk 10]
```diff
@@ -340,6 +340,7 @@ Other enhancements
 - :class:`pandas.core.window.ExponentialMovingWindow` now supports a ``times`` argument that allows ``mean`` to be calculated with observations spaced by the timestamps in ``times`` (:issue:`34839`)
 - :meth:`DataFrame.agg` and :meth:`Series.agg` now accept named aggregation for renaming the output columns/indexes. (:issue:`26513`)
 - ``compute.use_numba`` now exists as a configuration option that utilizes the numba engine when available (:issue:`33966`)
+- :meth:`DataFrame.agg` and :meth:`Series.agg` now accept named aggregation for renaming the output columns/indexes. (:issue:`26513`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:343]: is this a leftover?

**Review by jreback — CHANGES_REQUESTED** [review]
extra entry in the whats new

File: doc/source/whatsnew/v1.1.0.rst (hunk 10 above)
Comment (Steffen911) [reply doc/source/whatsnew/v1.1.0.rst]: @jreback I probably messed something up during merge conflict handling. Reverted this line to the documentation update I made in https://github.com/pandas-dev/pandas/pull/34514/commits/fa4915b7e39251b6ad6d19cdd6355009f94bce80

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green

**Comment by Steffen911** [comment]
@jreback All tests are greem. Travis reports itself as "Passed" when I go to the "Details", but apparently it didn't update the PR check after it has finished.

**Comment by TomAugspurger** [comment]
Thanks @Steffen911!

**Pull request merged**